en/tutorials-and-examples/blog/part-three.rst

```diff
@@ -0,0 +1,267 @@
+        }
+    }
+
+Generate Skeleton Code for Categories
+=====================================
+
+Create all files by launching bake commands::
+
+    bin/cake bake model Categories
+    bin/cake bake controller Categories
+    bin/cake bake view Categories
+
+The bake tool has created all your files in a snap. You can give them a quick
+read if you want re-familiarize yourself with how CakePHP works.
+
+Attach TreeBehavior to CategoriesTable
+======================================
+
+The :doc:`TreeBehavior </orm/behaviors/tree>` helps you manage hierarchical Tree
+structures in database table. It uses the `MPTT logic
+<http://www.sitepoint.com/hierarchical-data-database-2/>`_ to manage the data.
+MPTT tree structures are optimized for reads, which often makes them a good fit
+for read heavy applications like blogs.
+
+If you open the ``src/Model/Table/CategoriesTable.php`` file, you'll see
+that the TreeBehavior has been attached to your CategoriesTable in the
+``initialize`` method::
+
+    $this->addBehavior('Tree');
+
+With the TreeBehavior attached you'll be able to access some features like
+reordering the categories.  We'll see that in a moment.
+
+But for now, you have to remove the following inputs in your Categories add and
+edit template files::
+
+    echo $this->Form->input('lft');
+    echo $this->Form->input('rght');
+
+These fields are automatically managed by the TreeBehavior when
+a category is saved.
+
+Using your web browser, add some new categories using the
+``/yoursite/categories/add`` controller action.
+
+Reordering Categories with TreeBehavior
+========================================
+
+In your categories index template file, you can list the categories and ordering
+them.
+
+Let's modify the index method in your ``CategoriesController.php`` and add
+``move_up`` and ``move_down`` methods to be able to reorder the categories in
+the tree::
+
+    class CategoriesController extends AppController
+    {
+        public function index()
+        {
+            $categories = $this->Categories->find('threaded')
+                ->order(['lft' => 'ASC']);
+            $this->set(compact('categories'));
+        }
+
+        public function move_up($id = null)
+        {
+            $this->request->allowMethod(['post', 'put']);
+            $category = $this->Categories->get($id);
+            if ($this->Categories->moveUp($category)) {
+                $this->Flash->success('The category has been moved Up.');
+            } else {
+                $this->Flash->error('The category could not be moved up. Please, try again.');
+            }
+            return $this->redirect(['action' => 'index']);
+        }
+
+        public function move_down($id = null)
+        {
+            $this->request->allowMethod(['post', 'put']);
+            $category = $this->Categories->get($id);
+            if ($this->Categories->moveDown($category)) {
+                $this->Flash->success('The category has been moved down.');
```

Comment:
The code needs to redirect in the error case as well, because there is no moveup template. I'll add the referrer though.